website/docs/core-concepts/stacks/templates/functions/atmos.Vals.mdx

```diff
@@ -0,0 +1,155 @@
+```
+
+For more details, see: https://github.com/helmfile/vals?tab=readme-ov-file#expression-syntax
+
+Common expression patterns:
+
+- AWS Secret: `ref+awssecrets://myapp/dev/secret`
+- Vault Secret: `ref+vault://secret/data/myapp#/mykey`
+- SOPS: `ref+sops://path/to/file#/foo/bar`
+
+## Supported backends
+
+- [Vault](https://github.com/helmfile/vals#vault)
+- [AWS SSM Parameter Store](https://github.com/helmfile/vals#aws-ssm-parameter-store)
+- [AWS Secrets Manager](https://github.com/helmfile/vals#aws-secrets-manager)
+- [AWS S3](https://github.com/helmfile/vals#aws-s3)
+- [GCP Secrets Manager](https://github.com/helmfile/vals#gcp-secrets-manager)
+- [GCP KMS](https://github.com/helmfile/vals#gcp-kms)
+- [Google Sheets](https://github.com/helmfile/vals#google-sheets)
+- [Google GCS](https://github.com/helmfile/vals#google-gcs)
+- [SOPS](https://github.com/helmfile/vals#sops) powered by [sops](https://github.com/getsops/sops)
+- [Terraform (tfstate)](https://github.com/helmfile/vals#terraform-tfstate) powered by [tfstate-lookup](https://github.com/fujiwara/tfstate-lookup)
+- [Echo](https://github.com/helmfile/vals#echo)
+- [File](https://github.com/helmfile/vals#file)
+- [Azure Key Vault](https://github.com/helmfile/vals#azure-key-vault)
+- [EnvSubst](https://github.com/helmfile/vals#envsubst)
+- [GitLab](https://github.com/helmfile/vals#gitlab)
+- [1Password](https://github.com/helmfile/vals#1password)
+- [1Password Connect](https://github.com/helmfile/vals#1password-connect)
+- [Doppler](https://github.com/helmfile/vals#doppler)
+- [Pulumi State](https://github.com/helmfile/vals#pulumi-state)
+- [Kubernetes](https://github.com/helmfile/vals#kubernetes)
+- [Conjur](https://github.com/helmfile/vals#conjur)
+- [HCP Vault Secrets](https://github.com/helmfile/vals#hcp-vault-secrets)
+- [HTTP JSON](https://github.com/helmfile/vals#http-json)
+- [Bitwarden](https://github.com/helmfile/vals#bitwarden)
+
+For the complete list of supported backends, see: https://github.com/helmfile/vals?tab=readme-ov-file#supported-backends
+
+## Error Handling
+
+The `atmos.Vals` function will:
+- Return an error if the backend is unreachable
+- Fail fast if credentials are invalid
+- Surface backend-specific error messages
+
+Handle potential errors in your templates:
+
+```yaml
+# Basic error handling with default value
+{{ $secret := atmos.Vals "ref+vault://secret/myapp" | default "fallback-value" }}
+
+# Conditional logic based on secret availability
+{{- if $secret := atmos.Vals "ref+vault://secret/myapp" }}
+  value: {{ $secret }}
+{{- else }}
+  {{- fail "Required secret not available" }}
+{{- end }}
+
+# Multiple secrets with error aggregation
+{{- $creds := dict }}
+{{- range $key := list "username" "password" "api-key" }}
+  {{- $path := printf "ref+vault://secret/myapp/%s" $key }}
+  {{- $value := atmos.Vals $path | default (fail (printf "Missing required secret: %s" $key)) }}
+  {{- $_ := set $creds $key $value }}
+{{- end }}
+
```

Comment:
@aknysh this is only valid on template imports, right?

```suggestion
# Conditional logic based on secret availability
{{- if $secret := atmos.Vals "ref+vault://secret/myapp" }}
  value: {{ $secret }}
{{- else }}
  {{- fail "Required secret not available" }}
{{- end }}

# Multiple secrets with error aggregation
{{- $creds := dict }}
{{- range $key := list "username" "password" "api-key" }}
  {{- $path := printf "ref+vault://secret/myapp/%s" $key }}
  {{- $value := atmos.Vals $path | default (fail (printf "Missing required secret: %s" $key)) }}
  {{- $_ := set $creds $key $value }}
{{- end }}

```

Reply:
this example comes from coderabbit suggestion, i don't think you need something like this.

i'm using the function in many places simply like this:

```
components:
  terraform:
    argocd:
      vars:
        slack_token: '{{ atmos.Vals "ref+sops://secrets.enc.yaml#/argocd/slack_token" }}'
```

Reply:
btw switched to SOPS lately because remote backends like 1password can slows down command execution thats true